File: demo-assignment-management-system/server/src/services/canvasService.js
const axios = require('axios');
const { canvas } = require('../config/env');
const { ApiError } = require('../middleware/errorHandler');

/**
 * Thin wrapper around the Canvas LMS REST API.
 * Docs: https://canvas.instructure.com/doc/api/ (also linked from
 * https://instructure.github.io/)
 *
 * Auth model right now: a single access token from .env (fine for local
 * dev against your own Canvas account). Before this ships to real users,
 * swap to Canvas's OAuth2 authorization-code flow so each student
 * authorises their own account and we store per-user tokens instead of
 * one shared token.
 */
function client(accessToken) {
  const token = accessToken || canvas.accessToken;

  if (!canvas.baseUrl || !token) {
    throw new ApiError(
      500,
      'Canvas API is not configured. Set CANVAS_BASE_URL and CANVAS_ACCESS_TOKEN in server/.env'
    );
  }

  return axios.create({
    baseURL: `${canvas.baseUrl}/api/v1`,
    headers: { Authorization: `Bearer ${token}` },
    timeout: 10000,
  });
}

/** GET /courses — courses the current user is enrolled in */
async function getCourses(accessToken) {
  const { data } = await client(accessToken).get('/courses', {
    params: { enrollment_state: 'active', per_page: 100 },
  });
  return data;
}

/** GET /courses/:id/assignments — assignments for a single course */
async function getAssignmentsForCourse(courseId, accessToken) {
  const { data } = await client(accessToken).get(`/courses/${courseId}/assignments`, {
    params: { per_page: 100, order_by: 'due_at' },
  });
  return data;
}

/**
 * Fetch active courses, then fan out to fetch each course's assignments,
 * and flatten into one list. This is what feeds the "aggregate workload
 * across all papers" dashboard requirement.
 */
async function getAllUpcomingAssignments(accessToken) {
  const courses = await getCourses(accessToken);

  const assignmentLists = await Promise.all(
    courses.map((course) =>
      getAssignmentsForCourse(course.id, accessToken).catch(() => [])
      // a single course failing (e.g. no assignment permissions) shouldn't
      // take down the whole aggregate view
    )
  );

  return courses.map((course, i) => ({
    courseId: course.id,
    courseName: course.name,
    assignments: assignmentLists[i],
  }));
}

module.exports = { getCourses, getAssignmentsForCourse, getAllUpcomingAssignments };

File: demo-assignment-management-system/server/Services/CanvasService.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using AMS.Api.Config;
using AMS.Api.Middleware;
using AMS.Api.Models;
using Microsoft.Extensions.Options;

namespace AMS.Api.Services;

/// <summary>
/// Thin wrapper around the Canvas LMS REST API.
/// Docs: https://canvas.instructure.com/doc/api/ (also linked from
/// https://instructure.github.io/)
///
/// Registered as a typed HttpClient (see Program.cs), so a fresh HttpClient
/// is injected per instance and configured once here in the constructor.
///
/// Auth model right now: a single access token from configuration (fine for
/// local dev against your own Canvas account). Before this ships to real
/// users, swap to Canvas's OAuth2 authorization-code flow so each student
/// authorises their own account and we store per-user tokens instead of one
/// shared token.
/// </summary>
public class CanvasService : ICanvasService
{
    private readonly HttpClient _http;
    private readonly bool _configured;

    public CanvasService(HttpClient http, IOptions<CanvasOptions> options)
    {
        _http = http;
        var canvas = options.Value;
        _configured = canvas.IsConfigured;

        if (_configured)
        {
            _http.BaseAddress = new Uri($"{canvas.BaseUrl.TrimEnd('/')}/api/v1/");
            _http.DefaultRequestHeaders.Authorization =
                new AuthenticationHeaderValue("Bearer", canvas.AccessToken);
            _http.DefaultRequestHeaders.UserAgent.ParseAdd("AssignmentManagementSystem/1.0");
        }
    }

    private void EnsureConfigured()
    {
        if (!_configured)
        {
            throw new ApiException(500,
                "Canvas API is not configured. Set Canvas:BaseUrl and Canvas:AccessToken " +
                "(appsettings.json, env vars, or dotnet user-secrets).");
        }
    }

    public async Task<IReadOnlyList<CanvasCourse>> GetCoursesAsync()
    {
        EnsureConfigured();

        return await GetJsonOrThrowAsync<List<CanvasCourse>>(
            "courses?enrollment_state=active&per_page=100") ?? [];
    }

    public async Task<IReadOnlyList<CanvasAssignment>> GetAssignmentsForCourseAsync(long courseId)
    {
        EnsureConfigured();

        return await GetJsonOrThrowAsync<List<CanvasAssignment>>(
            $"courses/{courseId}/assignments?per_page=100&order_by=due_at") ?? [];
    }

    /// <summary>
    /// GetFromJsonAsync's built-in error handling throws away the response
    /// body on a non-success status, so all you ever see is "403 Forbidden"
    /// with no explanation. Canvas usually puts a specific reason in the
    /// body (e.g. "Invalid access token", "insufficient scope") — this reads
    /// it and puts it in the exception message instead of discarding it.
    /// </summary>
    private async Task<T?> GetJsonOrThrowAsync<T>(string requestUri)
    {
        var response = await _http.GetAsync(requestUri);

        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync();
            throw new ApiException((int)response.StatusCode,
                $"Canvas API request to '{requestUri}' failed with {(int)response.StatusCode} " +
                $"{response.ReasonPhrase}: {body}");
        }

        return await response.Content.ReadFromJsonAsync<T>();
    }

    /// <summary>
    /// Fetch active courses, then fan out to fetch each course's assignments
    /// in parallel, and flatten into one list. Feeds the "aggregate workload
    /// across all papers" dashboard requirement.
    /// </summary>
    public async Task<IReadOnlyList<CourseWithAssignments>> GetAllUpcomingAssignmentsAsync()
    {
        var courses = await GetCoursesAsync();

        var assignmentTasks = courses.Select(async course =>
        {
            try
            {
                return await GetAssignmentsForCourseAsync(course.Id);
            }
            catch
            {
                return (IReadOnlyList<CanvasAssignment>)[];
            }
        });

        var assignmentLists = await Task.WhenAll(assignmentTasks);

        return courses.Select((course, i) => new CourseWithAssignments
        {
            CourseId = course.Id,
            CourseName = course.Name,
            Assignments = assignmentLists[i],
        }).ToList();
    }
}